x/dfract/keeper/keeper.go

```diff
@@ -128,85 +152,61 @@ func (k Keeper) CreateDeposit(ctx sdk.Context, msg types.MsgDeposit) error {
-func (k Keeper) Mint(ctx sdk.Context, mintRate int64) error {
-	// Acquire the parameters to get the mint denom
-	params, err := k.GetParams(ctx)
-	if err != nil {
-		return err
-	}
-
-	// Compute the total amount
-	totalAmount := sdk.NewInt(0)
-	deposits := k.ListWaitingMintDeposits(ctx)
-	for _, deposit := range deposits {
-		totalAmount = totalAmount.Add(deposit.GetAmount().Amount)
-	}
+	// Acquire the list of waiting proposal & waiting mint deposits
+	waitingProposalDeposits := k.ListWaitingProposalDeposits(ctx)
+	waitingMintDeposits := k.ListWaitingMintDeposits(ctx)
 
-	// Make sure we actually have something to mint
-	if totalAmount.IsZero() {
-		return types.ErrMintDontMatchTotal
-	}
-
-	// Mint the coins
-	totalAmount.MulRaw(mintRate)
-	if err := k.MintKeeper.MintCoins(ctx, sdk.NewCoins(sdk.NewCoin(params.MintDenom, totalAmount))); err != nil {
-		return err
-	}
-	return nil
-}
-
-// Distribute Split the amount of emitted coins and distribute them to the depositors
-func (k Keeper) Distribute(ctx sdk.Context, mintRate int64, deposits []*types.Deposit) error {
-	// Acquire the parameters to get the mint denom
-	params, err := k.GetParams(ctx)
-	if err != nil {
-		return err
+	// Spend the coins from the module account if we have one
+	if balance.Amount.IsPositive() {
+		destinationAddress, err := sdk.AccAddressFromBech32(proposal.GetSpendDestination())
+		if err != nil {
+			return sdkerrors.ErrInvalidAddress
+		}
+		if err := k.BankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, destinationAddress, sdk.NewCoins(balance)); err != nil {
+			return err
+		}
 	}
 
-	// For each deposit, we will compute the amount of coins to send from the mint rate and process the distribution
-	for _, deposit := range deposits {
-		// Correctly format address
-		destinationAddress, err := sdk.AccAddressFromBech32(deposit.GetDepositorAddress())
+	// Process the waiting mint deposits
+	for _, deposit := range waitingMintDeposits {
+		depositorAddress, err := sdk.AccAddressFromBech32(deposit.GetDepositorAddress())
 		if err != nil {
 			return sdkerrors.ErrInvalidAddress
 		}
 
-		// Compute the coins to send from the rate
-		mintedAmount := sdk.NewCoin(params.MintDenom, deposit.GetAmount().Amount.MulRaw(mintRate))
+		toMint := sdk.NewCoin(params.MintDenom, deposit.GetAmount().Amount.MulRaw(proposal.GetMintRate()))
+		if err := k.BankKeeper.MintCoins(ctx, types.ModuleName, sdk.NewCoins(toMint)); err != nil {
+			return err
+		}
 
-		// Transfer the coins
-		if err := k.BankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, destinationAddress, sdk.NewCoins(mintedAmount)); err != nil {
+		if err := k.BankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, depositorAddress, sdk.NewCoins(toMint)); err != nil {
 			return err
 		}
 
-		// Remove from the waiting queue and move to minted
 		k.RemoveFromWaitingMintDeposits(ctx, deposit.GetDepositorAddress())
 		k.InsertIntoMintedDeposits(ctx, deposit.GetDepositorAddress(), *deposit)
```

Comment:
I wonder if this one might not trigger the same issue we had on the Beam module where the queue kept growing larger and larger which slowed down the network?